Debugging (#4608)
demisto/content: docs/integration_design/README.MD

# Integration Design
## Intro
In this tutorial we'll talk about how to get from an initial need or idea for an integration you want to implement in Demisto, to a design document that'll help you get organized and understand exactly what you need to implement.

## Why do I need a design document in the first place?
You might be thinking "Why do I need to write this document? Why not just start writing?" 

> Well, we beleive that writing such a document before starting any code writing will help you better understand your use of this integration, understand what exactly you want to do and what you want to get out of each command or fetch.

Reviewer: ```suggestion
Well, we believe that writing such a document before starting any code writing will help you better understand your use of this integration, understand what exactly you want to do and what you want to get out of each command or fetch.
```

9. [Network Security (IDS/IPS)](https://github.com/demisto/content/tree/master/docs/integration_design/usecases/network_security_firewall_usecase)
10. [Vulnerability Management](https://github.com/demisto/content/tree/master/docs/integration_design/usecases/vulnerability_management_usecase)

## Understanding and Documenting the Use Cases
If your'e on your way to writing an integration for Demisto you probably are already familiar with the product you want to integrate, but it might help to go into the product again to see what actions you want to implemnt, how you think they'll work once in Demisto.
Write down the use cases for your integration, how you would use it and how you think outher people will use it.
```Markdown
# Use Cases:
- Create, get, edit, close a ticket/issue, add + view comments.
- Assign a ticket/issue to a specified user.
- List all tickets, filter by name, date, assignee
```
## Parameters
In this section write what are the parameters you need in order to create an instance of the product and connect to it. These parameters will be accessible and used in all the integration. 

> We require 2 fields as part of the parameters:

Reviewer: mention common parameters that we see in integrations, like `url`, `apikey` / `apitoken`/ `credentials (username, password)`, if the integrations fetches incidents then `query` which will query events/incidents/tickets

1. proxy - You probably recoginze this field as "Use system proxy settings". This is a boolean field that handles proxy if True
2. insecure - (a.k.a. Trust any certificate (not secure)) This field is also boolean and indicates if to use only certified certificates.

When implementing the integration have a look at existing integrations to see how we implement this.
If your integration implements `Fetches incidents` also add a boolean field called `isFetch`
## Commands, Arguments & Outputs
Now let's write down exactly what commands you want to implement. Each command has 3 parts:
1. Command name
2. Command arguments
3. Command outputs
#### Command Name

> We have a standard for a command's name: `productName-command`

Reviewer: ```suggestion
We have a standard for a command's name: `brand-action`
```

For example `intsights-get-alert-activities`
If your commands checks the maliciousness of an IP/Domain/URL/File name your command simply `ip`\ `domain`\ `url`\ `file`. If this is the case make sure your outputs (which we'll get to in a second) match the [context standards](https://github.com/demisto/content/tree/master/docs/context_standards).

#### Command argument

> These are the arguments for the command for example `!intsights-get-alert-activities alert-id=`

Reviewer: ```suggestion
These are the arguments for the command for example `!intsights-get-alert-activities alert-id=`, `!ip ip="8.8.8.8"`
```

> These are the arguments for the command for example `!intsights-get-alert-activities alert-id=`

Reviewer: argument names must be snake case. 
good: `alert-id`.
bad: `alertId`

#### Command Outputs
There are 3 type of outputs from the command:
1. Human Readable output - This is what will be printed to the war room. It should be in Markdown form.
2. Raw Json - A JSON of the raw response that was returned from the API. 
3. Context Output - The outputs that will be in the context. 

The Context Outupt will be a JSON in the following format `{Product.TypeofOuput.Info}`. For more info about this have a look [here](https://github.com/demisto/content/tree/master/docs/context_and_ouputs)

> ### Command Deisgn

Reviewer: ```suggestion
### Command Design
```

The final result for each command should look something like this:

**Command Name**: cisco-asa-list-rules

###### Arguments
argument | type | description|mandatory
---|---|---|---

> InterfaceName|String|the name of the interface|no

Reviewer: InterfaceName is not according to our standard.

InterfaceType|choose|In/Out/Global|no

###### Context Outputs
* CiscoASA.Rule.Interface
* CiscoASA.Rule.SourceIP
* CiscoASA.Rule.DestIP
* CiscoASA.Rule.Permit
* CiscoASA.Rule.Active
* CiscoASA.Rule.Remarks
* CiscoASA.Rule.Position
* CiscoASA.Rule.ID
###### Human Readable Output
**All Rules for interface {} of type{}**:

RuleID|InterfaceName|SourceIP|DestIP|Position|
---|---|---|---|---|

> ...|...|...|...|...|

Reviewer: present some data here
